In [ ]:
!pip install pymupdf easyocr

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.0/25.0 MB 56.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 67.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 180.7/180.7 kB 13.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 978.2/978.2 kB 45.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 299.6/299.6 kB 21.1 MB/s eta 0:00:00


In [ ]:
import fitz  # PyMuPDF
import easyocr
import re

# 1. Initialize the OCR Reader (Industry standard: Load it once, not inside the function)
reader = easyocr.Reader(['en'])

def extract_text_from_pdf(pdf_path):
    """Extracts text from a digital (searchable) PDF."""
    doc = fitz.open(pdf_path)
    text = ""
    for page in doc:
        text += page.get_text()
    return text

def extract_text_from_image(image_path):
    """Extracts text from an image using Deep Learning."""
    results = reader.readtext(image_path, detail=0) # detail=0 gives just the text
    return " ".join(results)

def clean_text(text):
    """Industry Standard: Remove extra spaces and messy characters."""
    text = re.sub(r'\s+', ' ', text) # Replace multiple spaces/newlines with one space
    return text.strip()

Progress: |██████████████████████████████████████████████████| 100.0% Complete

Progress: |██████████████████████████████████████████████████| 100.0% Complete

In [ ]:
def extract_medical_values(text):
    patterns = {
        'glucose': r'[Gg]lucose[:\s]+(\d+\.?\d*)',
        'bp': r'[Bb]lood\s?[Pp]ressure[:\s]+(\d+/\d+)',
        'bmi': r'[Bb][Mm][Ii][:\s]+(\d+\.?\d*)'
    }

    extracted_data = {}
    for label, pattern in patterns.items():
        match = re.search(pattern, text)
        if match:
            extracted_data[label] = match.group(1)

    return extracted_data

In [ ]:
# 1. Simulate the messy text we got from a scanned report
raw_report_text = "HOSPITAL REPORT. Patient: Sameer. Findings: The Glucose: 155.2 mg/dL. Blood Pressure: 130/80. BMI is 28.4. End of report."

# 2. Clean it
cleaned_text = clean_text(raw_report_text)

# 3. Extract the structured data
medical_data = extract_medical_values(cleaned_text)

print("--- The AI's Structured Understanding ---")
print(medical_data)

--- The AI's Structured Understanding ---
{'glucose': '155.2', 'bp': '130/80'}


Lekin, kya aapne ek cheez notice ki? Hamare test text mein BMI bhi tha (BMI is 28.4), lekin output mein nahi aaya.
Kyun? (The Industry Lesson):
Kyuki hamara Regex sirf BMI: ya BMI (space) dhoond raha tha, lekin text mein BMI is likha tha.
The Reality: In the real world, doctors write in many ways: "Glucose level was 155", "Sugar: 155", "Glucose - 155".
Hum kitne Regex likhenge? 100? 1000? Nahi.
Isliye hum move karte hain Phase 6: NLP & Medical Entity Extraction (The "Deep Understanding" Brain).
